### PR TITLE
meta/sql: fix update dir stats by fault

### LINKED ISSUE #3725: fsck --sync-dir-stat --repair  can't work with sql meta engine.
**What happened**:

```
#!/bin/bash
set -ex
META_URL=sqlite3://test.db
umount /jfs || true
rm test.db -rf
rm /var/jfs/myjfs -rf
./juicefs format $META_URL myjfs
    ./juicefs mount -d $META_URL /jfs
    ./juicefs mdtest $META_URL /d --depth 10 --dirs 2 --files 2 --threads 10 &
    pid=$!
    sleep 15s
    kill -9 $pid
    ./juicefs info -r /jfs/d
    ./juicefs info -r /jfs/d --strict
    ./juicefs fsck $META_URL --path /d --sync-dir-stat --repair -r
    ./juicefs info -r /jfs/d | tee info1.log
    ./juicefs info -r /jfs/d --strict | tee info2.log
    diff info1.log info2.log
```

```
+ ./juicefs mdtest sqlite3://test.db /d --depth 10 --dirs 2 --files 2 --threads 10
2023/06/01 09:58:44.954871 juicefs[3858083] <INFO>: Meta address: sqlite3://test.db [interface.go:494]
2023/06/01 09:58:44.955862 juicefs[3858083] <INFO>: Data use file:///var/jfs/myjfs/ [mdtest.go:207]
2023/06/01 09:58:44.956022 juicefs[3858083] <INFO>: Disk cache (/var/jfsCache/7b9f812b-64aa-40e6-bd9f-b35e5bc288f0/): capacity (102400 MB), free ratio (10%), max pending pages (15) [disk_cache.go:114]
2023/06/01 09:58:44.962755 juicefs[3858083] <INFO>: Create session 2 OK with version: 1.1.0-dev+2023-05-31.ff6e765c [base.go:488]
2023/06/01 09:58:44.963223 juicefs[3858083] <INFO>: Prometheus metrics listening on 127.0.0.1:33101 [mount.go:160]
2023/06/01 09:58:44.963391 juicefs[3858083] <INFO>: Create 40940 files in 2047 dirs [mdtest.go:111]
2023/06/01 09:58:50.227646 juicefs[3858083] <INFO>: Created 2047 dirs in 5.264224411s (388 dirs/s) [mdtest.go:126]
2023/06/01 09:58:57.325395 juicefs[3858083] <WARNING>: Transaction succeeded after 3 tries (19.067717ms), inodes: [1], last error: database is locked [sql.go:722]
create file: 3807/40940 [=====>--------------------------------------------------------]  255.8/s ETA: 2m25s
+ kill -9 3858083
+ ./juicefs info -r /jfs/d
/jfs/d: 3691                        12137.0/s    
/jfs/d: 21.8 MiB (22835200 Bytes)   71.6 MiB/s   
/jfs/d :
  inode: 2
  files: 3691
   dirs: 2049
 length: 0 Bytes
   size: 21.78 MiB (22835200 Bytes)
   path: /d
./test.sh: line 13: 3858083 Killed                  ./juicefs mdtest $META_URL /d --depth 10 --dirs 2 --files 2 --threads 10
+ ./juicefs info -r /jfs/d --strict
/jfs/d: 3841                        12757.4/s    
/jfs/d: 23.0 MiB (24125440 Bytes)   76.4 MiB/s   
/jfs/d :
  inode: 2
  files: 3841
   dirs: 2049
 length: 0 Bytes
   size: 23.01 MiB (24125440 Bytes)
   path: /d
+ ./juicefs fsck sqlite3://test.db --path /d --sync-dir-stat --repair -r
2023/06/01 09:59:00.771759 juicefs[3858139] <INFO>: Meta address: sqlite3://test.db [interface.go:494]
+ ./juicefs info -r /jfs/d
+ tee info1.log
/jfs/d :
  inode: 2
  files: 1981
   dirs: 2049
 length: 0 Bytes
   size: 4.00 KiB (4096 Bytes)
   path: /d
+ ./juicefs info -r /jfs/d --strict
+ tee info2.log
/jfs/d :
  inode: 2
  files: 3841
   dirs: 2049
 length: 0 Bytes
   size: 23.01 MiB (24125440 Bytes)
   path: /d
+ diff info1.log info2.log
3c3
<   files: 1981
---
>   files: 3841
6c6
<    size: 4.00 KiB (4096 Bytes)
---
>    size: 23.01 MiB (24125440 Bytes)
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**

**Environment**:
- JuiceFS version (use `juicefs --version`) or Hadoop Java SDK version:
- Cloud provider or hardware configuration running JuiceFS:
- OS (e.g `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Object storage (cloud provider and region, or self maintained):
- Metadata engine info (version, cloud provider managed or self maintained):
- Network connectivity (JuiceFS to metadata engine, JuiceFS to object storage):
- Others:

### PR DESCRIPTION
**Pull request by Hexilee**
close #3725

we introduce a bug in #3715, which causes all dir stats will be updated by single record.

### CODE REVIEW

**Comment by zhoucheng361** [comment]
test passed.

**Review by zhoucheng361 — APPROVED** [review]

File: pkg/meta/sql.go [hunk 1]
```diff
@@ -2467,6 +2467,9 @@ func (m *dbMeta) doUpdateDirStat(ctx Context, batch map[Ino]dirStat) error {
 }
 
 func (m *dbMeta) doSyncDirStat(ctx Context, ino Ino) (*dirStat, syscall.Errno) {
+	if m.conf.ReadOnly {
```
Comment (SandyXSD) [inline pkg/meta/sql.go:2470]: add this as well for RedisMeta?

File: pkg/meta/sql.go [hunk 2]
```diff
@@ -2479,10 +2482,10 @@ func (m *dbMeta) doSyncDirStat(ctx Context, ino Ino) (*dirStat, syscall.Errno) {
 		if !exist {
 			return syscall.ENOENT
 		}
-		stats := &dirStats{ino, stat.length, stat.space, stat.inodes}
-		_, err = s.Insert(stats)
+		record := &dirStats{ino, stat.length, stat.space, stat.inodes}
+		_, err = s.Insert(record)
 		if err != nil && isDuplicateEntryErr(err) {
-			_, err = s.Cols("data_length", "used_space", "used_inodes").Update(stats)
+			_, err = s.Where("inode = ?", ino).Cols("data_length", "used_space", "used_inodes").Update(record)
```
Comment (SandyXSD) [inline pkg/meta/sql.go:2488]: usually we use `condiBean` in `Update` method.

Comment (Hexilee) [reply pkg/meta/sql.go]: done

File: pkg/meta/sql.go (hunk 1 above)
Comment (Hexilee) [reply pkg/meta/sql.go:2470]: done

**Pull request merged**